From cdd93775c38b0b5b9c12212e565c269ce544b7e0 Mon Sep 17 00:00:00 2001
From: Jai Luthra <j-luthra@ti.com>
Date: Tue, 9 May 2023 14:15:45 +0530
Subject: [PATCH 048/434] media: cadence: csi2rx: Propagate set_fmt from sink
 to source

We need to update the source format for the CSI-RX bridge when the user
sets sink format.

Signed-off-by: Jai Luthra <j-luthra@ti.com>
Signed-off-by: Vignesh Raghavendra <vigneshr@ti.com>
---
 drivers/media/platform/cadence/cdns-csi2rx.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/media/platform/cadence/cdns-csi2rx.c b/drivers/media/platform/cadence/cdns-csi2rx.c
index 222b78bc19b5..338f417b30db 100644
--- a/drivers/media/platform/cadence/cdns-csi2rx.c
+++ b/drivers/media/platform/cadence/cdns-csi2rx.c
@@ -480,7 +480,7 @@ static int csi2rx_set_fmt(struct v4l2_subdev *subdev,
 			  struct v4l2_subdev_format *format)
 {
 	struct csi2rx_priv *csi2rx = v4l2_subdev_to_csi2rx(subdev);
-	struct v4l2_mbus_framefmt *framefmt;
+	struct v4l2_mbus_framefmt *sinkfmt, *srcfmt;
 
 	/* No transcoding, source and sink formats must match. */
 	if (format->pad != CSI2RX_PAD_SINK)
@@ -491,13 +491,19 @@ static int csi2rx_set_fmt(struct v4l2_subdev *subdev,
 
 	format->format.field = V4L2_FIELD_NONE;
 
-	framefmt = csi2rx_get_pad_format(csi2rx, state, format->pad,
-					 format->which);
-	if (!framefmt)
+	sinkfmt = csi2rx_get_pad_format(csi2rx, state, format->pad,
+					format->which);
+	if (!sinkfmt)
+		return -EINVAL;
+
+	srcfmt = csi2rx_get_pad_format(csi2rx, state, CSI2RX_PAD_SOURCE_STREAM0,
+				       format->which);
+	if (!srcfmt)
 		return -EINVAL;
 
 	mutex_lock(&csi2rx->lock);
-	*framefmt = format->format;
+	*sinkfmt = format->format;
+	*srcfmt = format->format;
 	mutex_unlock(&csi2rx->lock);
 
 	return 0;
-- 
2.41.0

